metrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,302 @@
+package metrics
+
+import (
+	"github.com/prebid/prebid-cache/config"
+	"github.com/prometheus/client_golang/prometheus"
+	"time"
+)
+
+// Constants and global variables
+const (
+	StatusKey    string = "status"
+	FormatKey    string = "format"
+	ConnErrorKey string = "connection_error"
+
+	TotalsVal     string = "total"
+	ErrorVal      string = "error"
+	BadRequestVal string = "bad_request"
+	JsonVal       string = "json"
+	XmlVal        string = "xml"
+	DefinesTTLVal string = "defines_ttl"
+	InvFormatVal  string = "invalid_format"
+	CloseVal      string = "close"
+	AcceptVal     string = "accept"
+)
+
+//	Object definition
+type PrometheusMetrics struct {
+	Registry    *prometheus.Registry
+	Puts        *PrometheusRequestStatusMetric
+	Gets        *PrometheusRequestStatusMetric
+	PutsBackend *PrometheusRequestStatusMetricByFormat
+	GetsBackend *PrometheusRequestStatusMetric
+	Connections *PrometheusConnectionMetrics
+	ExtraTTL    *PrometheusExtraTTLMetrics
+}
+
+type PrometheusRequestStatusMetric struct {
+	Duration      prometheus.Histogram
+	RequestStatus *prometheus.CounterVec
+}
+
+type PrometheusRequestStatusMetricByFormat struct {
+	Duration           prometheus.Histogram
+	PutBackendRequests *prometheus.CounterVec
+	RequestLength      prometheus.Histogram
+}
+
+type PrometheusConnectionMetrics struct {
+	ConnectionsOpened prometheus.Gauge
+	ConnectionsErrors *prometheus.CounterVec
+}
+
+type PrometheusExtraTTLMetrics struct {
+	ExtraTTLSeconds prometheus.Histogram
+}
+
+//	Init functions
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	cacheWriteTimeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	requestSizeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1} // TODO: tweak
```

Comment:
Please resolve this TODO as we're getting close to merging. I think you'd be fine coverage a range up to 1MB, broken up into a max of 10 segments.

Reply:
done